File: backend/schemas/detection.py
"""
数据模型定义
"""
from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class DetectionBox(BaseModel):
    """检测框"""
    x1: float
    y1: float
    x2: float
    y2: float
    confidence: float
    class_id: int
    class_name: str


class DetectionResult(BaseModel):
    """单张图片检测结果"""
    image_name: str
    boxes: List[DetectionBox]
    class_counts: Dict[str, int]
    total_damages: int
    severity: str
    severity_score: float = 0.0  # 0~100 加权病害指数
    processing_time: float
    result_image: Optional[str] = None  # base64 编码的带框结果图
    record_id: Optional[int] = None  # 入库后的记录 ID
    is_duplicate: bool = False  # 是否命中去重（复用了已有检测结果）


class BatchDetectionResult(BaseModel):
    """批量检测结果"""
    total_images: int
    total_damages: int
    overall_severity: str
    overall_severity_score: float = 0.0
    class_counts: Dict[str, int]
    results: List[DetectionResult]


class VideoDetectionResult(BaseModel):
    """视频检测结果"""
    video_name: str
    total_frames: int
    processed_frames: int
    total_damages: int
    severity: str
    max_severity: str
    class_counts: Dict[str, int]
    fps: float
    duration: float
    result_video_url: Optional[str] = None


class TaskCreatedResponse(BaseModel):
    """异步任务创建响应"""
    task_id: str
    message: str = "任务已创建"


class TaskStatusResponse(BaseModel):
    """异步任务状态"""
    id: str
    status: str
    progress: float
    message: str
    total_frames: int
    processed_frames: int
    result: Optional[dict] = None
    error: Optional[str] = None


class DetectionConfig(BaseModel):
    """检测配置"""
    model_name: str = "高召回模型"
    conf_threshold: float = Field(default=0.5, ge=0.1, le=1.0)
    iou_threshold: float = Field(default=0.7, ge=0.1, le=1.0)
    selected_classes: Optional[List[str]] = None


class SystemStatus(BaseModel):
    """系统状态"""
    cpu_usage: float
    memory_usage: float
    gpu_usage: float
    gpu_memory: Optional[float] = None
    current_model: str
    available_models: List[str]


class RecordItem(BaseModel):
    """历史记录条目"""
    id: int
    detection_type: str
    source_name: str
    model_name: str
    conf_threshold: float
    iou_threshold: float
    total_damages: int
    class_counts: Dict[str, int]
    severity: str
    severity_score: float
    result_image_path: Optional[str] = None
    boxes: List[dict] = []
    processing_time: float
    created_at: Optional[str] = None
    extra: Optional[dict] = None


class RecordListResponse(BaseModel):
    """分页记录列表"""
    total: int
    page: int
    page_size: int
    items: List[RecordItem]

File: frontend/src/api/detection.ts
import request from './request'

export interface DetectionBox {
  x1: number
  y1: number
  x2: number
  y2: number
  confidence: number
  class_id: number
  class_name: string
}

export interface DetectionResult {
  image_name: string
  boxes: DetectionBox[]
  class_counts: Record<string, number>
  total_damages: number
  severity: string
  severity_score: number
  processing_time: number
  result_image?: string
  record_id?: number
  is_duplicate?: boolean
}

export interface BatchDetectionResult {
  total_images: number
  total_damages: number
  overall_severity: string
  overall_severity_score: number
  class_counts: Record<string, number>
  results: DetectionResult[]
}

export interface SystemStatus {
  cpu_usage: number
  memory_usage: number
  gpu_usage: number
  gpu_memory?: number
  current_model: string
  available_models: string[]
}

// 图像检测
export const detectImage = (formData: FormData) => {
  return request.post<any, DetectionResult>('/api/detection/image', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

// 批量检测
export const detectBatch = (formData: FormData) => {
  return request.post<any, BatchDetectionResult>('/api/detection/batch', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

// 获取系统状态
export const getSystemStatus = () => {
  return request.get<any, SystemStatus>('/api/system/status')
}

// 切换模型
export const switchModel = (modelName: string) => {
  return request.post(`/api/system/model/switch?model_name=${encodeURIComponent(modelName)}`)
}

// 获取可用模型
export const getAvailableModels = () => {
  return request.get<any, { models: string[], current: string }>('/api/system/models')
}
